### PR TITLE
JSON Web Token with Apache Shiro

### LINKED ISSUE #15: Authentication
How to authenticate users to our system

### PR DESCRIPTION
**Pull request by gjlam95**
Fix #15 #26

### CODE REVIEW

**Comment by Zhiyuan-Amos** [comment]
If you're comparing between solutions, do take a look at Spring Security also. Previously I was looking at Spring, I noticed that Spring Security does quite a lot of security stuffs too.

I have some hyperlinks in the tool assessment file.

File: tool_assessment.md [hunk 1]
```diff
@@ -45,27 +45,28 @@ There are 3 ways to ensure anonymity of data:
...
 
-Advanced Encryption Standard (AES) is a symmetric algorithm (private-key cryptography). This involves a single key which is a shared secret between the sender and recipient. The same key is being used for both encryption and decryption. Public-key cryptography (PKI), a asymmetric algorithm, involves two related keys for each recipient involved - a private key which is a secret known only by the recipient, and a related public key which is known by all senders. The sender encrypts the message using the recipient's public key. That message can only be decrypted by a recipient with a private key matching the public key. We will be using RSA for our asymmetric algorithm.
+The following diagram shows how a JWT is obtained and used to access APIs or resources:
 
-For simplicity and efficient encryption, we can employ the use of AES with a single secret key for encryption and decryption. The secret key used has to be stored in a safe location to prevent any potential compromise. We propose to lock it in a password protected .config file that outsiders are unable to access.
+![jwt_example](https://github.com/IFS4205-2018-Sem1-Team1/first-report/raw/master/images/jwt_example.png)
```
Comment (Zhiyuan-Amos) [inline tool_assessment.md]: Can't see this image. :P

File: tool_assessment.md [hunk 2]
```diff
@@ -45,27 +45,28 @@ There are 3 ways to ensure anonymity of data:
...
-## Encryption
+## Secure Data Transfer
+We intend to have a web application in which [authorization](https://en.wikipedia.org/wiki/Authorization) is handled by [Apache Shiro](https://shiro.apache.org/) and [authentication](https://en.wikipedia.org/wiki/Authentication) by [JSON Web Token (JWT)](https://jwt.io/). This would ensure secure data transfer when a user logs in with his/her own credentials.
 
-The purpose of [encryption](https://en.wikipedia.org/wiki/Encryption) is to ensure data transfer traffic is not susceptible to potential interceptors. Encryption will be used when data is transferred between the 3 different servers, namely front-end, back-end, and database.
+Apache Shiro is a powerful and easy to use Java security framework that offers developers an intuitive yet comprehensive solution to:
+1. Authentication
+1. Authorization
+1. Cryptography
+1. Session Management
```
Comment (Zhiyuan-Amos) [inline tool_assessment.md]: Are we using these features too? Cos things like Cryptography and Session management aren't mentioned later on.

If they aren't used, then there's no need to mention them cos it doesn't help in justifying when we are using this tool.

File: tool_assessment.md [hunk 3]
```diff
@@ -45,27 +45,28 @@ There are 3 ways to ensure anonymity of data:
...
 
-Example of encryption:
-![encryption-example](https://github.com/IFS4205-2018-Sem1-Team1/first-report/raw/master/images/public_key_encryption_keys.png)
+Although Apache Shiro already has a authorization functionality built into its library, we decide to employ the use of JWT for this purpose instead as JWT is much easier to implement as a Single Sign-On (SSO). In addition, it removes the need for cookies and session which is required in Apache Shiro's framework. Since JWT only provides authorization, we would still require Apache Shiro for authentication.
```
Comment (Zhiyuan-Amos) [inline tool_assessment.md]: Explanation looks good. Just one question, what's single sign on? XD should we elaborate more here?

File: tool_assessment.md [hunk 4]
```diff
@@ -45,27 +45,28 @@ There are 3 ways to ensure anonymity of data:
...
+The following diagram shows how a JWT is obtained and used to access APIs or resources:
 
-For simplicity and efficient encryption, we can employ the use of AES with a single secret key for encryption and decryption. The secret key used has to be stored in a safe location to prevent any potential compromise. We propose to lock it in a password protected .config file that outsiders are unable to access.
+![jwt_example](https://github.com/IFS4205-2018-Sem1-Team1/first-report/raw/master/images/jwt_example.png)
 
-Possible library alternatives:
-1. [PHP](http://php.net/manual/en/refs.crypto.php)
-1. [Python](https://docs.python.org/3/library/crypto.html)
-1. [Perl](https://perldoc.perl.org/functions/crypt.html)
+A justification for the use of JWT is because it comes with a huge [library](https://jwt.io/#libraries) supported for many different programming languages.
```
Comment (Zhiyuan-Amos) [inline tool_assessment.md]: I think this justification isn't very strong cos we are only using for a particular language right? :P as in it's not like we are using Jwt for sql, php, java and so on.

File: tool_assessment.md [hunk 5]
```diff
@@ -45,27 +45,28 @@ There are 3 ways to ensure anonymity of data:
...
+When we use Apache Shiro to implement the login system, a JWT information is returned to the front-end, and it stores the token information in the request header when it interacts with the back-end server. We will then configure a custom interceptor to intercept all URL requests, retrieve the token information in the request header information, and verify the token information. If the token information generated during the login is correct, the user is logged in. Otherwise, we will reject the request and return a 401 error.
+
+An alternative to Apache Shiro is [Spring Security](https://spring.io/projects/spring-security). While it is not dependent on running in a Servlet/JSP environment, it is heavily biased towards web applications in terms of ease of developer use.  Apache Shiro and Spring Security differ mainly on scope and mental model/design. Apache Shiro has a broader scope than Spring Security as it also addresses problems associated with enterprise session management as well as cryptography, concurrency, etc. Shiro also has one 'killer feature' that people generally reference a lot: built-in support for very fine-grained (e.g. instance-level) access control.
```
Comment (Zhiyuan-Amos) [inline tool_assessment.md]: > While it is not dependent on running in a Servlet/JSP environment, it is heavily biased towards web applications in terms of ease of developer use.

I don't understand this :(

**Comment by Zhiyuan-Amos** [comment]
Your PR fixes #15 too. Add it to your header so that it will auto close the issue when this PR is merged

**Comment by gjlam95** [comment]
Changed the justification but unsure if it's better :(

File: tool_assessment.md (hunk 5 above)
Comment (gjlam95) [reply tool_assessment.md]: hmm maybe i should remove the first phrase, kinda not relevant for us also

File: tool_assessment.md [hunk 6]
```diff
@@ -45,27 +45,26 @@ There are 3 ways to ensure anonymity of data:
...
+Apache Shiro is a powerful and easy to use Java security framework that offers developers an intuitive yet comprehensive solution to:
+1. Authentication
+1. Authorization
 
-Example of encryption:
-![encryption-example](https://github.com/IFS4205-2018-Sem1-Team1/first-report/raw/master/images/public_key_encryption_keys.png)
+Although Apache Shiro already has a authorization functionality built into its library, we decide to employ the use of JWT for this purpose instead as JWT is much easier to implement as a Single Sign-On (SSO). Single sign-on (SSO) is a property of access control of multiple related, yet independent, software systems. With this property, a user logs in with a single ID and password to gain access to a connected system or systems without using different usernames or passwords, or in some configurations seamlessly sign on at each system. In addition, it removes the need for cookies and session which is required in Apache Shiro's framework. Since JWT only provides authorization, we would still require Apache Shiro for authentication.
```
Comment (Zhiyuan-Amos) [inline tool_assessment.md]: > authorization is handled by Apache Shiro

> Although Apache Shiro already has a authorization functionality... we decide to employ the use of JWT for this purpose instead

Bo tally haha :o

Comment (Zhiyuan-Amos) [inline tool_assessment.md]: I'm not so clear about the `Single Sign-On`, cos I thought we are only having 1 software system. Cos I was thinking of SSO for logging in to different government agencies (which are different software systems), but in our case we only have 1 right?

File: tool_assessment.md [hunk 7]
```diff
@@ -45,27 +45,26 @@ There are 3 ways to ensure anonymity of data:
...
+![jwt_example](https://github.com/IFS4205-2018-Sem1-Team1/first-report/raw/master/images/jwt_example.png)
 
-Possible library alternatives:
-1. [PHP](http://php.net/manual/en/refs.crypto.php)
-1. [Python](https://docs.python.org/3/library/crypto.html)
-1. [Perl](https://perldoc.perl.org/functions/crypt.html)
+A justification for the use of JWT is because it comes with an extensive [library](https://jwt.io/#libraries) for Java programming.
 
-AES [example](https://aesencryption.net/).
+When we use Apache Shiro to implement the login system, a JWT information is returned to the front-end, and it stores the token information in the request header when it interacts with the back-end server. We will then configure a custom interceptor to intercept all URL requests, retrieve the token information in the request header information, and verify the token information. If the token information generated during the login is correct, the user is logged in. Otherwise, we will reject the request and return a 401 error.
```
Comment (Zhiyuan-Amos) [inline tool_assessment.md]: Should this paragraph be shifted to be directly below the image? (Is this paragraph is the description of the image?)

File: tool_assessment.md [hunk 8]
```diff
@@ -45,27 +45,26 @@ There are 3 ways to ensure anonymity of data:
...
+
+An alternative to Apache Shiro is [Spring Security](https://spring.io/projects/spring-security). Apache Shiro and Spring Security differ mainly on scope and mental model/design. Apache Shiro has a broader scope than Spring Security as it also addresses problems associated with enterprise session management. Shiro also has one 'killer feature' that people generally reference a lot: built-in support for very fine-grained (e.g. instance-level) access control.
```
Comment (Zhiyuan-Amos) [inline tool_assessment.md]: This segment have to elaborate on how it's relevant to our project. Generally, the different frameworks all have their pros and cons, so we need to show how Spring's cons may eventually cause some issues with our project, which will not happen if we use Apache Shiro.

File: tool_assessment.md [hunk 9]
```diff
@@ -64,7 +64,7 @@ When we use Apache Shiro to implement the login system, a JWT information is ret
 
 A justification for the use of JWT is because it comes with an extensive [library](https://jwt.io/#libraries) for Java programming.
 
-An alternative to Apache Shiro is [Spring Security](https://spring.io/projects/spring-security). Apache Shiro and Spring Security differ mainly on scope and mental model/design. Apache Shiro has a broader scope than Spring Security as it also addresses problems associated with enterprise session management. Shiro also has one 'killer feature' that people generally reference a lot: built-in support for very fine-grained (e.g. instance-level) access control. 
+An alternative to Apache Shiro is [Spring Security](https://spring.io/projects/spring-security). However, for our case, Apache Shiro requires very few dependencies, making it lightweight. In contrast, Spring Security is known for being rather heavyweight and bound to Spring tech in some cases. When we have to handle a large amount of user queries, Apache Shiro would be best suited to handle the server load.
```
Comment (Zhiyuan-Amos) [inline tool_assessment.md]: > bound to Spring tech in some cases. 

The downside of bounding to Spring tech isn't clear. Is it cos there's more dependencies to other parts of Spring tech, which causes it to be heavyweight?

> When we have to handle a large amount of user queries, Apache Shiro would be best suited to handle the server load.

Does the previous point cause this issue? :P From what I know, it's just having bigger file sizes (cos you need to compile & build the dependencies also). So the argument can be, both do the job equally well, but Apache Shiro depends on fewer dependencies, making it lightweight. That's why we're using Apache Shiro.

File: tool_assessment.md [hunk 10]
```diff
@@ -45,27 +45,26 @@ There are 3 ways to ensure anonymity of data:
...
+![jwt_example](https://github.com/IFS4205-2018-Sem1-Team1/first-report/raw/master/images/jwt_example.png)
 
-Possible library alternatives:
-1. [PHP](http://php.net/manual/en/refs.crypto.php)
-1. [Python](https://docs.python.org/3/library/crypto.html)
-1. [Perl](https://perldoc.perl.org/functions/crypt.html)
+When we use Apache Shiro to implement the login system, a JWT information is returned to the front-end, and it stores the token information in the request header when it interacts with the back-end server. We will then configure a custom interceptor to intercept all URL requests, retrieve the token information in the request header information, and verify the token information. If the token information generated during the login is correct, the user is logged in. Otherwise, we will reject the request and return a 401 error.
 
-AES [example](https://aesencryption.net/).
+A justification for the use of JWT is because it comes with an extensive [library](https://jwt.io/#libraries) for Java programming.
```
Comment (Zhiyuan-Amos) [inline tool_assessment.md]: I think this paragraph should be merged with line 55, or should be placed after line 55. Cos you're discussing about why we're using JWT over there.

**Pull request merged**